pandas/core/apply.py

```diff
@@ -397,100 +382,73 @@ def agg_dict_like(self) -> DataFrame | Series:
         """
         return self.agg_or_apply_dict_like(op_name="agg")
 
-    def agg_or_apply_dict_like(
-        self, op_name: Literal["agg", "apply"]
-    ) -> DataFrame | Series:
-        from pandas import Index
-        from pandas.core.groupby.generic import (
-            DataFrameGroupBy,
-            SeriesGroupBy,
-        )
-        from pandas.core.reshape.concat import concat
-
-        assert op_name in ["agg", "apply"]
-
+    def compute_dict_like(
+        self,
+        op_name: Literal["agg", "apply"],
+        selected_obj: Series | DataFrame,
+        selection: Hashable | Sequence[Hashable],
+        kwargs: dict[str, Any],
+    ) -> tuple[list[Hashable], list[Any]]:
```

Comment:
I assume it's `list[Any]` here because  `result_data` may be anything if `obj` is `SeriesGroupby` or `Series` while its `list[Series]` if `obj` is a `DataFrame` or `DataFrameGroupby` object?

Could you explain in the doc string a bit about when it returns different return types?

Reply:
I don't feel I can with any certainty. This can be used with an arbitrary UDF combined with the fact that this is (at times) used by Series, DataFrame, GroupBy, Window, and Resampler. I _think_ SeriesGroupBy will always return Series (but not certain).

I added docstrings with an indication of at least one case where arbitrary Python objects can definitely be returned.

Reply:
Ok, I understand.